CDP4Composition/Services/FavoritesService/FavoritesService.cs

```diff
@@ -0,0 +1,216 @@
+                    {
+                        var userPreference = (UserPreference)objectChange.ChangedThing;
+                        return userPreference.Container == session.ActivePerson &&
+                               userPreference.ShortName == this.GetPreferenceShortname(type);
+                    })
+                    .Select(o => this.GetIidsFromUserPreference((UserPreference)o.ChangedThing))
+                    .ObserveOn(RxApp.MainThreadScheduler)
+                    .Subscribe(subscriberAction);
+        }
+
+        /// <summary>
+        /// Toggles the favorite status of a thing.
+        /// </summary>
+        /// <typeparam name="T">The type on which to base the save.</typeparam>
+        /// <param name="thing">The thing to persist/remove from persistence</param>
+        /// <param name="session">The session in which to persist.</param>
+        /// <returns>The empty task.</returns>
+        public async Task ToggleFavorite<T>(ISession session, T thing) where T: Thing
+        {
+            var transaction = new ThingTransaction(TransactionContextResolver.ResolveContext(session.ActivePerson));
+            
+            var preferenceShortname = this.GetPreferenceShortname(typeof(T));
+            var preference = this.GetFavoritePreference(session, preferenceShortname);
+
+            var userClone = session.ActivePerson.Clone(false);
+
+            HashSet<Guid> valueSet;
+
+            if (preference == null)
+            {
+                // if property not there, create it and add
+                valueSet = new HashSet<Guid>
+                {
+                    thing.Iid
+                };
+
+                preference = new UserPreference(Guid.NewGuid(), null, null)
+                {
+                    ShortName = preferenceShortname
+                };
+
+                userClone.UserPreference.Add(preference);
+                transaction.CreateOrUpdate(userClone);
+            }
+            else
+            {
+                // if property is there, see if thing is in the array, true => remove, false => append
+                valueSet = this.GetIidsFromUserPreference(preference);
+
+                if (valueSet.Contains(thing.Iid))
+                {
+                    valueSet.Remove(thing.Iid);
+                }
+                else
+                {
+                    valueSet.Add(thing.Iid);
+                }
+
+                preference = preference.Clone(false);
+            }
+
+            preference.Value = string.Join(",", valueSet);
+            transaction.CreateOrUpdate(preference);
+
+            try
+            {
+                var operationContainer = transaction.FinalizeTransaction();
+                await session.Write(operationContainer);
+            }
+            catch (Exception ex)
+            {
+                LogManager.GetLogger(typeof(FavoritesService).FullName).Error("The inline update operation failed: {0}", ex.Message);
```

Comment:
logger is instantiated once on top of the class, no need to get it for every exception